### DISCUSSION TITLE
Privacy Primitives

### DISCUSSION THREAD

**Discussion by Christewart** [bitcoin-s/bitcoin-s#2513] [Ideas]
Trying to learn how CoinSwaps actually work for bitdevs tonight, and this made be think if we can start adding in "Privacy Primitives" into `core` / `wallet` to make it easier to implement things like this for the bitcoin-s. 

https://en.bitcoin.it/wiki/CoinSwap 

https://gist.github.com/chris-belcher/9144bd57a91c194e332fb5ca371d0964

If we were to do something like this, we should probably add it on as a new module that depends on the wallet. 

It would seem wise to be me (unlike we currently do with the [DLCWallet](https://github.com/bitcoin-s/bitcoin-s/blob/d8b0c21f2011c85eae13bd5905e400cf87144f4d/dlc-wallet/src/main/scala/org/bitcoins/dlc/wallet/DLCWallet.scala#L37)) to have a _has-a_ relationship between our wallet project rather _is-a_ wallet. 

Benefits would be 

1. Decouple privacy implementations from boring data storage stuff in the wallet
2. Forces us to make our core wallet apis more flexible to support a wide variety of use cases.

**Comment by benthecarman** [comment]
Adding tor support would be critical for this as well.

**Reply by Christewart** [reply]
> Adding tor support would be critical for this as well.

For some reference, @rorp actually has already added Tor support to eclair. I would assume it wouldn't be _extremely_ difficult to pull that code over here and put it in a bitcoin-s module 

https://github.com/ACINQ/eclair/pull/736

**Comment by Tv-igtmomsp** [comment]
There Definitely Will Be Over Time Better And Better Wallets!! Like Voice Activated Wallets For Those That Have Special Abilities and Aren't Able To See💯AI is what's up!!👌🏾
